src/components/App.tsx

```diff
@@ -4261,6 +4260,41 @@ class App extends React.Component<AppProps, AppState> {
       if (this.state.selectedLinearElement) {
         const linearElementEditor =
           this.state.editingLinearElement || this.state.selectedLinearElement;
+
+        const ret = LinearElementEditor.addMidpoint(
+          this.state.selectedLinearElement,
+          pointerCoords,
+          this.state,
+        );
+
+        if (
+          ret?.pointerDownState.segmentMidpoint.value !== null &&
+          !ret?.pointerDownState.segmentMidpoint.added
+        ) {
+          return;
+        }
+
+        if (ret?.didAddPoint) {
+          this.setState({
+            selectedLinearElement: {
+              ...this.state.selectedLinearElement,
+              pointerDownState: ret.pointerDownState,
+              selectedPointsIndices: ret.selectedPointsIndices,
+            },
+          });
+
+          if (this.state.editingLinearElement) {
+            this.setState({
+              editingLinearElement: {
+                ...this.state.editingLinearElement,
+                pointerDownState: ret.pointerDownState,
+                selectedPointsIndices: ret.selectedPointsIndices,
+              },
+            });
+          }
+          return;
+        }
```

Comment:
The `LinearElementEditor.pointerDownState.segmentMidpoint` should be part of global `pointerDownState`, or possibly the whole `LinearElementEditor.pointerDownState` should be. Because if we're updating `LinearElementEditor.pointerDownState` in a way that requires it to be saved back to `appState` (instead of mutating it), there's a race condition unless we `flushSync` it everywhere we update the linear editor.

The whole flow we're handling state during pointerdown/pointermove is super brittle and subject to race conditions right now. This manifests in things like:

![excal-midpoint-creation-loop](https://user-images.githubusercontent.com/5153846/204149460-2bdd463c-5bf3-48e2-a68b-628427330f82.gif)

Wrapping these calls into `flushSync` seems to fix it, but there are many other places where we do not set (linear editor) state synchronously, so those places may be subject to race conditions as well.

Reply:
Hey, can you share the exact steps to reproduce this? For me adding midpoints inside the editor works fine.

As per mutation yes we are using `flushSync` in a few places where the updated state needs to be used but though it's a hacky fix, we will need to rewrite it properly so we can take advantage of batching

Reply:
Oh, forgot to mention, but it may not work for you even with these steps. The idea is (I think) to introduce enough lag into the pointermouse handler so that the state doesn't update before the next pointermove handler fires.

For what, all I needed to do is:
1) add a console.log after the `addMidPoint()` call in `App.tsx`
2) enter line editor
3) drag to add midpoint

You can try adding more logs if one doesn't work for you.